File: server_py/abi/providers/llm/ollama.py
"""Ucretsiz/local alternatif. Gemini yoksa otomatik devreye girer."""

from __future__ import annotations

import json
from typing import AsyncIterator

import httpx

from ...config import config
from .base import LlmRequest


class OllamaProvider:
    id = "ollama"
    label = "Ollama (local)"
    vision = False

    async def available(self) -> bool:
        try:
            async with httpx.AsyncClient(timeout=0.7) as client:
                response = await client.get(f"{config.ollama.base_url}/api/tags")
                return response.status_code == 200
        except httpx.HTTPError:
            return False

    async def stream(self, request: LlmRequest) -> AsyncIterator[str]:
        messages = [{"role": "system", "content": request.system}]
        messages += [
            {"role": "user" if turn.role == "user" else "assistant", "content": turn.text}
            for turn in request.history
        ]
        messages.append({"role": "user", "content": request.message})

        body = {
            "model": config.ollama.model,
            "messages": messages,
            "stream": True,
            "options": {"temperature": request.temperature, "num_predict": request.max_tokens},
        }

        async with httpx.AsyncClient(timeout=httpx.Timeout(120.0, connect=5.0)) as client:
            async with client.stream("POST", f"{config.ollama.base_url}/api/chat", json=body) as response:
                response.raise_for_status()
                async for line in response.aiter_lines():
                    if not line.strip():
                        continue
                    try:
                        chunk = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    text = (chunk.get("message") or {}).get("content")
                    if text:
                        yield text

File: server/src/providers/llm/ollama.ts
import { config } from '../../config.js';
import type { LlmProvider, LlmRequest } from '../types.js';

/** Ucretsiz/local alternatif. Gemini yoksa otomatik devreye girer. */
export class OllamaProvider implements LlmProvider {
  readonly id = 'ollama';
  readonly label = 'Ollama (local)';

  async available(): Promise<boolean> {
    try {
      const res = await fetch(`${config.ollama.baseUrl}/api/tags`, {
        signal: AbortSignal.timeout(700),
      });
      return res.ok;
    } catch {
      return false;
    }
  }

  async *stream(req: LlmRequest): AsyncIterable<string> {
    const messages = [
      { role: 'system', content: req.system },
      ...req.history.map((t) => ({
        role: t.role === 'user' ? 'user' : 'assistant',
        content: t.text,
      })),
      { role: 'user', content: req.message },
    ];

    const res = await fetch(`${config.ollama.baseUrl}/api/chat`, {
      method: 'POST',
      headers: { 'content-type': 'application/json' },
      signal: req.signal,
      body: JSON.stringify({
        model: config.ollama.model,
        messages,
        stream: true,
        options: { temperature: req.temperature, num_predict: req.maxTokens },
      }),
    });
    if (!res.ok || !res.body) throw new Error(`ollama: ${res.status} ${res.statusText}`);

    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buffer = '';
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let nl: number;
      while ((nl = buffer.indexOf('\n')) !== -1) {
        const line = buffer.slice(0, nl).trim();
        buffer = buffer.slice(nl + 1);
        if (!line) continue;
        try {
          const parsed = JSON.parse(line) as { message?: { content?: string } };
          const text = parsed.message?.content;
          if (text) yield text;
        } catch {
          /* yarim satir */
        }
      }
    }
  }
}
